Improvement mark modified fields as dirty

SamyMP opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.5`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.5` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Completes #13222

Specifically resolves [this discussion](https://github.com/pimcore/pimcore/pull/13222#issuecomment-1434305241).

## Additional info  

Transforms setters from this :
```php
public function setMyField($myField)
{
    $this->myField = $myField;

    return $this;
}
```
to this :
```php
public function setMyField($myField)
{
    if ($this->getMyField() !== $myField) {
        $this->markFieldDirty('myField', true);
        
        $this->myField = $myField;
    }
    return $this;
}
```

Which then allows a user to know if a field has been modified or not in an EventListener by implementing [DirtyIndicatorInterface](https://github.com/pimcore/pimcore/blob/11.x/models/Element/DirtyIndicatorInterface.php) and use trait [DirtyIndicatorTrait](https://github.com/pimcore/pimcore/blob/11.x/models/Element/Traits/DirtyIndicatorTrait.php)

To keep track of which field's value has changed, a small loop in the DataObjectController checks the last known value of each field in the database against the value of the current object's state. 

What do you think ?

SamyMP committed:
Marks modified fields as dirty in model setters

SamyMP committed:
Mark fields that have changed as dirty in DataObjectController

jdreesen commented on models/DataObject/ClassDefinition/Data.php:
```diff
@@ -552,7 +552,10 @@ public function getSetterCode(DataObject\Objectbrick\Definition|DataObject\Class
             $code .= "\t" . '$this->' . $key . ' = ' . '$fd->preSetData($this, $' . $key . ');' . "\n";
         } else {
-            $code .= "\t" . '$this->' . $key . ' = ' . '$' . $key . ";\n\n";
+            $code .= "\t" . 'if($this->get' . ucfirst($this->getName()) . '() !== $' . $key . ') {' . "\n";
```
```suggestion
            $code .= "\t" . 'if ($this->get' . ucfirst($this->getName()) . '() !== $' . $key . ') {' . "\n";
```

SamyMP committed:
Add space after if instruction

Co-authored-by: Jacob Dreesen <jacob@hdreesen.de>

SamyMP committed:
fix tests

SamyMP committed:
fix tests

SamyMP commented:
Why are tests failing ? 🙂

jdreesen commented:
> PHP Fatal error:  Allowed memory size of 1073741824 bytes exhausted (tried to allocate 262144 bytes) in /home/runner/work/pimcore/pimcore/tests/_output/var/classes/DataObject/Csstore.php on line 53

SamyMP commented:
So the code takes too much memory to execute ? 🤔

mcop1 commented:
> So the code takes too much memory to execute ? 🤔

Exactly, it is running out of memory. I had a quick look and there seems to be a recursion now: 
![grafik](https://user-images.githubusercontent.com/89011527/221796391-ea70a96c-a3bf-405a-974c-18d2b2ac957a.png)

`setCsstore` now calls `getCsstore` (to compare the values) which then again calls `setCsstore` and so on ... 
Cold you please have a look?

SamyMP commented:
Hi, thank you for your input. I did try to execute the tests locally to see what's wrong but I couldn't make it work.

I'll try to work on that when I have time

SamyMP committed:
fix infinite loop caused by ClassificationStore

mcop1 commented:
@SamyMP ready for review again, right?

SamyMP commented:
@mcop1 Yes it is, sorry I didn't warn you

mcop1 commented on tests/Model/DataObject/ClassDefinitionTest.php:
```diff
@@ -66,8 +66,11 @@ public function testInputSetterCode(): void
 public function setInput(?string $input): static
 {
-	$this->input = $input;
+	if ($this->getInput() !== $input) {
+		$this->markFieldDirty("input", true);
 
+		$this->input = $input;
+	}
```
Do we still need those changes, now that you have fixed the test in another way?

mcop1 commented on tests/Model/DataObject/ClassDefinitionTest.php:
```diff
@@ -134,8 +137,11 @@ public function testQuantityValueSetterCode(): void
 public function setQuantityValue(?\Pimcore\Model\DataObject\Data\QuantityValue $quantityValue): static
 {
-	$this->quantityValue = $quantityValue;
+	if ($this->getQuantityValue() !== $quantityValue) {
+		$this->markFieldDirty("quantityValue", true);
 
+		$this->quantityValue = $quantityValue;
+	}
```
Same here

mcop1 commented on tests/Model/DataObject/ClassDefinitionTest.php:
```diff
@@ -156,8 +162,11 @@ public function testLocalizedFieldSetterCode(): void
 public function setLinput(?string $linput): static
 {
-	$this->linput = $linput;
+	if ($this->getLinput() !== $linput) {
+		$this->markFieldDirty("linput", true);
 
+		$this->linput = $linput;
+	}
```
And here

SamyMP committed:
Merge branch '11.x' of github.com:SamyMP/pimcore into improvement-mark-modified-fields-as-dirty

SamyMP committed:
Remove unecessary getters in setter

SamyMP commented:
@mcop1 I resolved the conversations, I modified the code accordingly to my comment. The `get` call in the setter was not necessary as it is how it's done today. I just added the `markFieldDirty` method in the setter, it is the focus of this PR

mcop1 commented:
Thank you :)

kingjia90 commented on bundles/AdminBundle/src/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -1385,10 +1385,19 @@ public function saveAction(Request $request): JsonResponse
+                $fields = array_keys($object->getClass()->getFieldDefinitions());
+                foreach ($fields as $field) {
+                    $getter  ='get' . ucfirst($field);
+                    if ($object->$getter() !== $objectFromDatabase->$getter()) {
```
I am facing a problem where the object and arrays are always seen as changed 🤔suspected that the 3 equal could be faulty when the array key order is different, but still doesn't work as intended. The only fix is to convert to array and use `array_diff_assoc`, any ideas?

To reproduce, try to save a new `Customer`, and put a breakpoint in there, it would trigger that `Calculated Segments` has changed.
